Docker image documentation (#1053)
earthly/earthly: docs/docker-images/all-in-one.md

This image contains `earthly`, `buildkit`, and some extra configuration to enable the two to work together. All that's missing is your source code! This image is mainly intended for use in containerized CI scenarios, or where maintaining a persistent installation of `earthly` isn't possible.

## Quickstart

Want to just get started? Here are a couple sample `docker run` commands that cover the most common use-cases:

### Simple Usage

```bash

> $ docker run --privileged -t -e NO_DOCKER=1 -v $(pwd):/workspace earthly/earthly:latest +for-linux

Reviewer: Mind removing the `$`? We've not used `$` in other snippets, to allow the user to copy-paste easier.

Author: Preventing blind copy paste was why I did it ;) I can remove it if you would like. These CLIs will almost always require tweaking.

```

Heres a quick breakdown:

> - `--privileged` is required when you are using the internal `buildkit`. This is because `buildkit` currently requires is when used with `earthly`.

Reviewer: ```suggestion
- `--privileged` is required when you are using the internal `buildkit`. This is because `buildkit` currently requires it for overlayfs support.
```

- `-t` tells Docker to emulate a TTY. This makes the `earthly` log output colorized.
- `-e NO_DOCKER=1` skips the check for a functional Docker daemon. If you are not exporting any images, you do not need Docker.
- `-v $(pwd):/workspace` mounts the source code into the conventional location within the docker container. Earthly is executed from this directory when starting the container. Any artifacts saved within this folder remain on your local machine.
- `+for-linux` is the target to be invoked. All arguments specified after the image tag will be passed to `earthly`.

### More Complicated Usage

```bash

> $ docker run -t --privileged -v $(pwd):/workspace:rw --network=host -v ~/.earthly/config.yml:/etc/.earthly/config.yml -e DOCKER_HOST="tcp://0.0.0.0:2375" earthly/earthly:corey_entrypoint --ci -P +for-linux

Reviewer: `$`

> $ docker run -t --privileged -v $(pwd):/workspace:rw --network=host -v ~/.earthly/config.yml:/etc/.earthly/config.yml -e DOCKER_HOST="tcp://0.0.0.0:2375" earthly/earthly:corey_entrypoint --ci -P +for-linux

Reviewer: It's more common to use `/var/run/docker.sock` for Docker. Would that be a better example?

Author: Yes; but why would you when you are already running privileged? The earlier example works locally; this was intended to show you could use an external `DOCKER_HOST`, as you might in a CI.

Author: I tweaked the address and wording a bit to make this less host-centric, at the expense of the command not working out of the box for a user.

Reviewer: > Yes; but why would you when you are already running privileged?

One reason is that Docker automatically creates `/var/run/docker.sock`, but it doesn't automatically listen on TCP.

I think both are good examples. Perhaps the table mentioning DOCKER_HOST could mention both examples.

Author: I guess what I was trying to get at is that if you are already privileged, isn't best practice to run your own daemon? I'll add more details in the table if that works for you rather than adding another quickstart with `-v /var/run/docker.sock:/var/run/docker.sock`?

Reviewer: Yeah that makes sense

| DOCKER_HOST                         | `/var/run/docker.sock`         | From Docker's CLI.                                                                                                                                                                                    |
| BUILDKIT_HOST                       | `tcp://<hostname>:8372`        | The address of your buildkit host. Use this when you have a remote `buildkitd` you would like to connect to.                                                                                          |             
| SRC_DIR                             | `/workspace`                   | The working directory for `earthly`. Usually host-mounted.                                                                                                                                            |
| EARTHLY_ADDITIONAL_BUILDKIT_CONFIG  |                                | Additional `buildkitd` config to append to the generated configuration file.                                                                                                                          |

> | BUILDKIT_LOCAL_REGISTRY_LISTEN_PORT | `8371`                         | What port should the internal cache registry listen on?                                                                                                                                               |

Reviewer: I think this should be disabled for both images. The reason is that Earthly code assumes the registry is at `127.0.0.1:8371` and doesn't support anything else. If it's enabled, buildkit will try to export via registry and earthly would not be able to receive the images, leading to failures.

Author: I haven't touched this setting at all; this was just documenting what was there. You want me to actively force it to the default value of 8371?

Reviewer: Perhaps don't document its existence just yet. The default should be unset, in which case buildkit will not use a registry for exports.

Author: It remains unset in here. The defaults in this table arent always from my script; but rather are what they end up defaulting to if unspecified... eventually. I will remove this from the config.

| CNI_MTU                             | MTU of first default interface | Set this when we autodetect the MTU incorrectly. The device used for autodetection can be shown by the command  `ip route show \| grep default \| cut -d' ' -f5 \| head -n 1`                         |
| EARTHLY_RESET_TMP_DIR               | `false`                        | Cleans out `EARTHLY_TMP_DIR` before running, if set to `true`. Useful when you host-mount an temp dir across runs.                                                                                    |
| NETWORK_MODE                        | `cni`                          | Specifies the networking mode of `buildkitd`. Default uses a CNI bridge network, configured with the `CNI_MTU`.                                                                                       |
| EARTHLY_TMP_DIR                     | `/tmp/earthly`                 | Specifies the location of `earthly`s temp dir. You can also mount an external volume to this path to preserve the contents across runs.                                                               |

> | CACHE_SIZE_MB                       | `0`                            | How big should the `buildkitd` cache be allowed to get, in MiB? 0 is unbounded.                                                                                                                       |

Reviewer: ```suggestion
| CACHE_SIZE_MB                       | `0`                            | How big should the `buildkitd` cache be allowed to get, in MiB? A value of 0 sets the cache size to "adaptive", causing buildkit to detect the available size of the system and choose a limit automatically.                                                                                                                       |
```
